### ISSUE TITLE
Issue with Layers on Windows Phone?

### ISSUE THREAD

**Issue by JohnnyWestlake** [Microsoft/Win2D#50]
I was wondering is there are any know issues with D2D layers in general on Windows Phone? (other than it being not-too-performance friendly)

I've attached a video of my issue ( as I'm having trouble getting a slim repo project at the moment), but basically sometimes if I use layers, my rendering breaks, and it also causes glitches to the rendering of the XAML parts of my UI as well. (In the video, the bar at the bottom is *not* a Windows Phone app bar, but just a normal Stackpanel / grid combo. You'll see it's rendering flashing all over the screen. If I take the layer call out, everything works perfectly all of the time :/

This is using a CanvasAnimatedControl inside a XAML page on Windows Phone 8.1 Update (Lumia 1520 with Black update). Also happens on a Lumia 635

Now, this black background / repeated texture also only happens if the last layer I created was the one that for some reason failed (although in the case of this app, the rendering code for each layer is exactly the same, as I'm rendering the same thing in different positions, using layers for optional opacity). If I create another layer call after this (which you see in the video when I add another copy), the screen renders fine again apart from the fact that broken layer simply doesn't render at all.

It looks like it might be a problem with the driver or a bug in D2D in general? (It's worth pointing out I've seen similar things happen not-too-infrequently in Internet Explorer on lower powered devices, although never on the one I recorded the video on - Lumia 1520)

Video: https://onedrive.live.com/redir?resid=DE47921B27CEAD6E!185016&authkey=!AELNCgKcIo-nFN4&ithint=video%2cmp4

**Comment by shawnhar** [comment]
Thanks for the report!   This does look like it could be a driver a bug.  I will do some digging and talk to the D2D team about it.

**Comment by JohnnyWestlake** [comment]
So I spent a lot of time trying to create my own simple repo, but it turns out I didn't really need too! The Win2D Example gallery actually demonstrates a similar (and I assume likely related issue) on it's layer examples page.

Switch though the layer modes one by one on a phone until you get to the last example (clip geometry). Watch random glitchyness ensue. Now, the problem seems to get exascerbated when there's more to draw, so changing the XAML a little bit like below (to get the canvas to stretch most of the screen on the phone), and then doubling the amount of items in the bouncingball array in the code behind (just pasting the elements again at the end of the array) makes the problem get worse.



```

    <Grid>
        <Grid.ColumnDefinitions>
            <ColumnDefinition Width="Auto"/>
            <ColumnDefinition Width="*"/>
        </Grid.ColumnDefinitions>
        
              <canvas:CanvasAnimatedControl Grid.ColumnSpan="2" 
                                      ClearColor="CornflowerBlue"
                                      CreateResources="CreateResources"
                                      Update="Update"
                                      Draw="Draw"/>

        <StackPanel Orientation="Vertical" MinWidth="160" Margin="5" HorizontalAlignment="Left" Background="Transparent">
            <TextBlock>Mode:</TextBlock>
            <ComboBox ItemsSource="{Binding ElementName=control, Path=ModeList}"
                      SelectedItem="{Binding CurrentMode, ElementName=control, Mode=TwoWay}"/>
        </StackPanel>

  
    </Grid>
```


Tested on a Lumia 1020, Lumia 635 and Lumia 1520. The problem seems to be less evident on the 1020 but still obviously there, scaling up to the rendering being almost completely broken on the 1520

**Comment by JohnnyWestlake** [comment]
Furthermore, it seems whatever is actually causing this is fixed in the latest public preview of Windows Phone 10. 

Testing side by side with a Lumia 635 on Windows Phone 10 and a Lumia 635 on the latest public Windows Phone 8.1 version gives very different results - the 8.1 version is a glitch mess whereas the 10 preview version is fine.

(Unless it's a driver issue, and the 10.0 preview is using some generic driver whereas the shipping phones don't? Although at this point at I little idea what I'm talking about :P)

**Comment by shawnhar** [comment]
Thanks for the extra digging!   I've also managed to repro this issue, and it does seem to be a driver problem.  I opened a bug on the D2D team and will see what we can do.  Unfortunately driver level issues are quite time consuming to get to the bottom of, so we won't be able to turn this one around as quickly as we normally could with a bug in Win2D itself.

**Comment by shawnhar** [comment]
Update: we are making progress toward a solution for this in Windows 10.   I will update again once that is definite.

**Comment by JohnnyWestlake** [comment]
Awesome, any chance you can give us a little insight into what the underlying problem is? (Not only because it might help give people some ides come up with a workaround / know what to avoid, but technical stuff is generally interesting to read even if I don't really understand what's going on all the time :P)

**Comment by shawnhar** [comment]
It's looking like a driver bug, but I don't understand all the details yet.

**Comment by shawnhar** [comment]
Final update on this:

It is indeed a driver bug.  We'll make sure this gets fixed in newer drivers, and for devices with older drivers, have added a workaround to the Windows 10 OS that will avoid hitting the particular case that isn't working correctly.

The release path for getting new preview OS builds out onto developer phones is fairly long, but this will definitely be fixed in Win10 RTM.

Unfortunately, we were not able to find a reasonable way to work around the issue purely in the Win2D layer, so the fix will only be available in Win10, not 8.1.
